### PR TITLE
Intake bits

### PR DESCRIPTION
**Pull request by vichik123**
This pull request simply adds the retractor bit.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+public class RetractorTest extends CommandBase {
+    private final Timer timer = new Timer();
+    private final Intake intake;
+    private final int numOfRuns;
+    private double previousTimeStamp = 0;
+    private boolean previousState;
+    private int counter = 0;
+
+    public RetractorTest(int numOfRuns, Intake intake) {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java]:
```suggestion
    public RetractorTest(Intake intake, int numOfRuns) {
```

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    public void initialize() {
+        timer.reset();
+        timer.start();
+        previousState = intake.getRetractorState();
+    }
+
+    @Override
+    public void execute() {
+        double currentTimeStamp = timer.get();
+        if (currentTimeStamp - previousTimeStamp > Constants.Intake.TIME_BETWEEN_RUNS) {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java]: I think (but not sure) you can use `timer.advanceIfElapsed()` and make it easier

File: src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -82,6 +82,7 @@ public static TalonFXConfiguration getConfiguration() {
         public static final double DEFAULT_POWER = 0.5; // power intake will receive on the basic command. [%]
         public static final double POWER_TO_VELOCITY_RATIO = -3 / 16.0; // Ratio of power to velocity. [% / m/s]
         public static final boolean IS_COMPENSATING_VOLTAGE = true;
+        public static final double TIME_BETWEEN_RUNS = 1.0; // time intake will wait before toggling the retractor (for testing only)
```
Comment (adamv17) [inline src/main/java/frc/robot/Constants.java]: Add in seconds

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        this.numOfRuns = numOfRuns;
+        this.intake = intake;
+        addRequirements(intake);
+    }
+
+    @Override
+    public void initialize() {
+        timer.reset();
+        timer.start();
+        previousState = intake.getRetractorState();
```
Comment (adamv17) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java:28]: Add first toggle call here

**Review by adamv17 — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        this.intake = intake;
+        addRequirements(intake);
+    }
+
+    @Override
+    public void initialize() {
+        timer.reset();
+        timer.start();
+
+        intake.toggleRetractor();
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java:27]: I don't think you should toggle the retractor in this method

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        if (timer.advanceIfElapsed(Constants.Intake.TIME_BETWEEN_RUNS)) {
+            intake.toggleRetractor();
+            boolean currentState = intake.getRetractorState();
+
+            if (previousState != currentState)
+                System.out.println("run number " + counter + " worked");
+
+            previousState = currentState;
+            counter++;
+            timer.reset();
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java:42]: You don't need to call reset, advanceIfElapsed already does it

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        }
+    }
+
+    @Override
+    public void end(boolean interrupted) {
+        intake.closeRetractor();
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java:48]: Stop the timer

File: src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public void execute() {
+        if (timer.advanceIfElapsed(Constants.Intake.TIME_BETWEEN_RUNS)) {
+            intake.toggleRetractor();
+            boolean currentState = intake.getRetractorState();
+
+            if (previousState != currentState)
+                System.out.println("run number " + counter + " worked");
+
+            previousState = currentState;
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/intake/commands/bits/RetractorTest.java:40]: You don't need to check, the value will always change (if there is a retractor)

**Review by barel13 — APPROVED** [review]

**Review by barel13 — CHANGES_REQUESTED** [review]
Oops

**Pull request merged**